extensions/search.md

```diff
@@ -0,0 +1,130 @@
+---
+
+## Notes for implementing work-in-progress version
+
+This is a work-in-progress specification.
+
+Software implementing this work-in-progress specification MUST NOT use the
+unprefixed `search` capability name. Instead, implementations SHOULD
+use the `draft/search` capability name to be interoperable with other
+software implementing a compatible work-in-progress version.
+
+The final version of the specification will use an unprefixed capability name.
+
+## Introduction
+
+This feature is intended to clients to search server-side through messages that were previously sent. Server-side search enables clients to quickly find messages rather than downloading all the history from their server.
+
+It is *not* a goal of this specification to offer context around matching messages, as this is covered by the [`chathistory`](chathistory.md) specification.
+
+The server as mentioned in this document may refer to either an IRC server or an IRC bouncer.
+
+Full support for this extension requires support for the batch, server-time and message-tags capabilities. However, limited functionality is available to clients without support for these CAPs. Servers SHOULD NOT enforce that clients support all related capabilities before using the search extension.
+
+## Architecture
+
+### Capabilities
+
+This specification introduces the `search` capability. This capability advertises to clients that the `SEARCH` command is available.
+
+The `search` capability MUST be negotiated.
+
+### Batch type
+
+This specification introduces the `search` batch type. This batch type is used when listing matching messages of a `SEARCH` command.
+
+### Messages
+
+This specification introduces the `SEARCH` commmand.
+
+## Descriptiob
+
+Clients can request a message search by sending the `SEARCH` command to the server.
+
+This command has the following general syntax:
+
+    SEARCH <attributes>
+
+The `attributes` parameter is a dictionary of attributes keys and values, formatted according to the [`message-tags`](message-tags.md) specification (without the leading `@`).
+
+If the batch capability was negotiated, the server MUST reply to a successful SEARCH command using a batch with batch type `search`. If no content exists to return, the server SHOULD return an empty batch in order to avoid the client waiting for a reply.
+
+The server then replies with a batch of batch type `search` containing messages matching all the specified match attributes. These messages MUST be `PRIVMSG` or `NOTICE` messages.
+
+### Returned message notes
+
+The order of returned messages within the batch is implementation-defined, but SHOULD be ascending time order or some approximation thereof, regardless of the subcommand used. The server-time tag on each message SHOULD be the time at which the message was received by the IRC server. When provided, the msgid tag that identifies each individual message in a response MUST be the msgid tag as originally sent by the IRC server.
```

Comment:
What about supporting specification of a specific ordering to return? (Oldest, newest, most relevant, etc)